File: sushicore/sushicore/FastFail.swift
//
//  FastFail.swift
//  sushicore
//
//  Created by Michael Ong on 28/6/17.
//  Copyright © 2017 Michael Ong. All rights reserved.
//

import Foundation

public struct   FastFailType: Error { public var code, message: String }

extension       FastFailType: Equatable {

    public static func  ==              (lhs: FastFailType, rhs: FastFailType) -> Bool { return lhs.code == rhs.code }
}


public protocol FastFailDelegate: class {

    func                programCrashed  (with failureType: FastFailType)
}


public class    FastFail {

    public private(set)
    static var          instance        : FastFail = FastFail()


    public weak var     delegate        : FastFailDelegate?

    public func         crash           (with failureType: FastFailType) -> Never {

        delegate?.programCrashed(with: failureType)

        fatalError("PROGRAM FAILURE: \(failureType.code) - \(failureType.message)")
    }

    public func         nonFatal        (with failureType: FastFailType) -> Void {

        print("NON-FATAL: \(failureType.code) - \(failureType.message)")
    }
}


extension       FastFailType {

    public static var unimplemented         = FastFailType(code: "xx_00", message: "Unimplemented method, please implement me!")

    public static var unexpectedPayload     = FastFailType(code: "po_01", message: "Invalid payload specified.")

    public static var driverNotInstalled    = FastFailType(code: "du_00", message: "Driver not installed!")
}

File: sushicpp/sushicore/sushicore/impl/FastFail.cpp
// FastFail.cpp
// sushicore - sushicpp
// 
// Created: 00:50 08/15/2017
// Author: Michael Ong

#include "../FastFail.h"

#include <string>
#include <iostream>

using namespace sushi::core;

using namespace std;

FastFail*   FastFail::instance          = instance = new FastFail();


FastFail::FastFail                      () {

}

FastFail::~FastFail                     () {

}

// ReSharper disable once CppMemberFunctionMayBeStatic
void        FastFail::crash             (const exception error) const {

    cerr    << "💥 Program Crashed with exception "
            << error.what()
            << endl;

    abort();
}

// ReSharper disable once CppMemberFunctionMayBeStatic
void        FastFail::nonFatal          (const exception error) const {

    cerr    << "💥 Program non-fatal error with exception "
            << error.what()
            << endl;
}

using namespace failureTypes;



driverNotInstalled::driverNotInstalled(const string &what, const string &fileName, const int &lineNumber):

    runtime_error("DRIVER_NOT_INSTALLED: " + fileName + "-" + to_string(lineNumber) + ": " + what ) {

}


unexpectedPayload::unexpectedPayload(const string &what, const string &fileName, const int &lineNumber):

    runtime_error("UNEXPECTED_PAYLOAD: " + fileName + "-" + to_string(lineNumber) + ": " + what ) {

}

unimplemented::unimplemented(const string &what, const string &fileName, const int &lineNumber):

    runtime_error("UNIMPLEMENTED: " +  fileName + "-" + to_string(lineNumber) + ": " + what ) {

}
